Feature/dot quote

HerbertGao opened the pull request:
## Description:
- Added a native `dot://` notification plugin that supports both the Dot Text API (`/text/`) and Image API (`/image/`).
- The plugin communicates with Quote/0 hardware devices via the Dot. service API (the devices are marketed as "Quote/0" but use the "Dot." API branding).
- Introduced explicit URL modes, payload routing, and field validation that mirror Dot's official API schema.
- **Attachment support**: The plugin supports file attachments that are automatically converted to base64:
  - **Text mode**: First attachment is used as `icon` (40×40 PNG) if no `icon` parameter is provided
  - **Image mode**: First attachment is used as `image` (296×152 PNG) if no `image` parameter is provided
  - Only the first attachment is used; multiple attachments trigger a warning
- Documented new syntax and examples in `README.md`, plus added regression coverage via `tests/test_plugin_dot.py`.

## Dot. Notifications

* **Source**: https://dot.mindreset.tech
* **Icon Support**: Yes (Text API accepts a 40×40 PNG icon via base64)
* **Message Format**: Plain Text
* **Message Limit**: Not documented

**Note on Device Naming**: The Dot. service API is used to communicate with hardware devices marketed as **Quote/0**. While the physical devices are sold under the "Quote/0" brand name, the API service and mobile application are branded as "Dot." This plugin uses the `dot://` protocol scheme to match the official API branding, ensuring consistency with Dot.'s documentation and developer tools.

The Dot. smart display exposes two REST endpoints—Text API and Image API—for pushing short reminders or graphics. The new plugin mirrors these entry points while keeping the URL grammar intuitive.

### 🛠️ Setup Instructions

1. Open the Dot. mobile app and retrieve both your **API token** (`dot_app_...`) and device **serial number** (12-character hex string).
2. In the app, enable the **Text API** and/or **Image API** content slot for the device.
3. Use the token and device ID with the `dot://` URLs shown below to trigger notifications.

---

### ✅ Apprise Support

### Syntax

- **Text API**  
  `dot://{token}@{device_id}/text/?signature={footer}&icon={base64_icon}`
  
  Title and message can be provided at runtime via `apprise.notify(title="...", body="...")`.

- **Image API**  
  `dot://{token}@{device_id}/image/?image={base64_png}&link={tap_url}&border={0|1}&dither_type={type}&dither_kernel={kernel}`

### Attachment Support

The plugin supports file attachments that are automatically converted to base64 encoding:

- **Text mode**: The first attachment is automatically used as the `icon` parameter (40×40 PNG) if no `icon` is provided via URL or configuration
- **Image mode**: The first attachment is automatically used as the `image` parameter (296×152 PNG) if no `image` is provided via URL or configuration
- If multiple attachments are provided, only the first one is used and a warning is logged
- If `icon` (text mode) or `image` (image mode) is already provided via URL/configuration, attachments are ignored

---

### 🔐 Parameter Breakdown

| Variable     | Required    | Description                                                                                                                         |
| ------------ | ----------- | ----------------------------------------------------------------------------------------------------------------------------------- |
| token        | Yes         | Dot. API token (`dot_app_...`)                                                                                                      |
| device_id    | Yes         | Dot. device serial number (12 hex characters)                                                                                       |
| title        | No (text)   | Title shown on device                                                                                                               |
| message      | No (text)   | Body text shown on device                                                                                                           |
| signature    | No (text)   | Footer text shown on device                                                                                                         |
| icon         | No (text)   | Base64 PNG icon (40×40) for the lower-left corner. Can be provided via URL parameter or first attachment (auto-converted to base64) |
| image        | Yes (image) | Base64 PNG image (296×152) rendered full-screen. Can be provided via URL parameter or first attachment (auto-converted to base64)   |
| link         | No          | Tap-to-interact target (http/https or custom scheme)                                                                                |
| border       | No (image)  | 0=white (default), 1=black frame                                                                                                    |
| ditherType   | No (image)  | DIFFUSION, ORDERED, or NONE                                                                                                         |
| ditherKernel | No (image)  | THRESHOLD, ATKINSON, BURKES, FLOYD_STEINBERG, SIERRA2, STUCKI, JARVIS_JUDICE_NINKE, DIFFUSION_ROW, DIFFUSION_COLUMN, DIFFUSION_2D   |

---

### 📦 Examples

**Send a text reminder (via URL parameters):**

```bash
apprise -vv -t "Morning Routine" -b "Remember to water the plants" \
  dot://dot_app_TOKEN@A1B2C3D4E5F6/text/?signature=Apprise
```

**Send a text reminder with icon via attachment:**

```bash
apprise -vv -t "Morning Routine" -b "Remember to water the plants" \
  -a /path/to/icon.png \
  dot://dot_app_TOKEN@A1B2C3D4E5F6/text/?signature=Apprise
```

**Push an image card (via URL parameter):**

```bash
apprise -vv \
  dot://dot_app_TOKEN@A1B2C3D4E5F6/image/?image=$(base64 -w0 poster.png)&link=https://example.com
```

**Push an image card via attachment:**

```bash
apprise -vv -a /path/to/image.png \
  dot://dot_app_TOKEN@A1B2C3D4E5F6/image/?link=https://example.com
```

**Python API usage with attachments:**

```python
import apprise

# Text mode with attachment as icon
apobj = apprise.Apprise()
apobj.add('dot://dot_app_TOKEN@A1B2C3D4E5F6/text/')
apobj.notify(
    title='Notification Title',
    body='Notification Body',
    attach=['/path/to/icon.png']  # Auto-converted to base64 as icon
)

# Image mode with attachment as image
apobj = apprise.Apprise()
apobj.add('dot://dot_app_TOKEN@A1B2C3D4E5F6/image/')
apobj.notify(
    attach=['/path/to/image.png']  # Auto-converted to base64 as image
)
```

## New Service Completion Status

* [x] apprise/plugins/dot.py
* [x] pypackage.toml update `keywords` section to identify our new service
    - add new service into this file (alphabetically).
* [x] README.md
    - add entry for new service to table (as a quick reference)
* [x] packaging/redhat/python-apprise.spec
    - add new service into the `%global common_description`

## Checklist

* [x] The code change is tested and works locally.
* [x] There is no commented out code in this PR.
* [x] No lint errors (tox -e lint / tox -e format run clean).
* [x] Test coverage added (`tests/test_plugin_dot.py`).

## Testing

Anyone can verify the plugin with:

```bash
# Optional fresh environment
python3 -m venv apprise-demo
cd apprise-demo
source bin/activate

pip install git+https://github.com/HerbertGao/apprise.git@feature/dot-quote

# Text API smoke test
apprise -vv -t "Sample Title" -b "Sample Body" \
  dot://dot_app_TOKEN@DEVICE_ID/text/?signature=Tester

# Image API smoke test (via URL parameter)
apprise -vv dot://dot_app_TOKEN@DEVICE_ID/image/?image=<base64png>

# Image API smoke test (via attachment)
apprise -vv -a /path/to/image.png dot://dot_app_TOKEN@DEVICE_ID/image/

# Text API with icon attachment
apprise -vv -t "Test" -b "Body" -a /path/to/icon.png \
  dot://dot_app_TOKEN@DEVICE_ID/text/

# Full tox test suite (requires repo checkout)
tox -e apprise -- -t "Sample Title" -b "Sample Body" \
  dot://dot_app_TOKEN@DEVICE_ID/text/
```

caronc commented on apprise/plugins/dot.py:
```diff
@@ -0,0 +1,636 @@
+    # Define object templates
+    templates = (
+        "{schema}://{apikey}@{device_id}/{mode}",
+        "{schema}://{apikey}@{device_id}/{mode}/",
```
Duplicate line; one above is fine

caronc commented on apprise/plugins/dot.py:
```diff
@@ -0,0 +1,636 @@
+            "message": {
+                "name": _("Text Content"),
+                "type": "string",
+            },
```
Not sure if I get this. I think I can follow the default title (fallback if one isn't specified), but the body is usually mandatory

caronc commented on apprise/plugins/dot.py:
```diff
@@ -0,0 +1,636 @@
+        message_value = results["qsd"].get("message")
+        if message_value:
+            results["message"] = NotifyDot.unquote(message_value.strip())
+
```
Drop 🙏🙂

HerbertGao commented on pyproject.toml:
```diff
@@ -134,6 +135,7 @@ keywords = [
     "PushSafer",
     "Pushy",
     "QQ Push",
+    "Quote/0",
```
“Quote/0” is the name of the device currently supported by Dot. Any content sent through Dot will be displayed on the Quote/0 device.

HerbertGao commented:
This submission addresses all code review feedback. The changes align the Dot plugin implementation with the Apprise framework standards.

## Main Changes

- Removed duplicate `protocol = "dot"` declaration  
- Removed duplicate template definitions  
- Removed default title/message logic and ensured title and body are only passed as runtime parameters  
  - Since Dot’s text mode allows empty titles, it no longer falls back to `app_desc`  
- Added warning logs when title or body is provided in image mode, informing users these parameters will be ignored  
- Enabled attachment support (`attachment_support = True`)  
  - Text mode: the first attachment is automatically converted to base64 and used as the `icon` parameter (40 px × 40 px PNG)  
  - Image mode: the first attachment is automatically converted to base64 and used as the `image` parameter (296 px × 152 px PNG)  
  - Only the first attachment is used; a warning is logged if multiple attachments are provided  
  - If `icon` (text mode) or `image_data` (image mode) is already provided via config parameters, attachments are ignored  
  - Attachment processing errors are logged but do not interrupt notification sending (text mode) or cause failures (image mode)  
- Simplified parameter validation logic  
  - Removed default-value validation and conversion for `border`, `dither_type`, and `dither_kernel`  
  - These parameters are now passed directly to the API, which handles defaults  
  - Parameters are only included in the request payload when not `None`

## Breaking Changes

Previously supported URLs (now invalid):

- `dot://apikey@device/text/?title=DefaultTitle&message=DefaultMsg`

Correct usage (provide title/body at runtime):

```python
apprise_obj = Apprise()

apprise_obj.add('dot://apikey@device/text/')

apprise_obj.notify(
    title='Dynamic Title',    # Provided at send time
    body='Dynamic Content'    # Provided at send time
)
```

## Attachment Usage Examples

**Text mode – use attachment as icon:**

```python
apprise_obj = Apprise()
apprise_obj.add('dot://apikey@device/text/')

# The attachment is automatically converted to base64 and used as icon
apprise_obj.notify(
    title='Notification Title',
    body='Notification Content',
    attach=['/path/to/icon.png']  # 40x40 PNG
)
```

**Image mode – use attachment as image:**

```python
apprise_obj = Apprise()
apprise_obj.add('dot://apikey@device/image/')

# The attachment is automatically converted to base64 and used as image
apprise_obj.notify(
    attach=['/path/to/image.png']  # 296x152 PNG
)
```

**Note:** If `icon` (text mode) or `image` (image mode) is already provided through URL parameters or configuration, the attachment will be ignored.

caronc commented on README.md:
```diff
@@ -28,24 +28,27 @@ System Administrators and DevOps who wish to send a notification now no longer n
-  * [File Attachments](#api-file-attachments)
-  * [Loading Custom Notifications/Hooks](#api-loading-custom-notificationshooks)
-* [Persistent Storage](#persistent-storage)
-* [More Supported Links and Documentation](#want-to-learn-more)
```
Only thing I have left is to restore all of this, the only change made to the readme should be the 'dot' entry further down.

Otherwise I can't spot anything else. Great work. This is very close to a merge and is a full service. Very impressive 🚀
